### PR TITLE
[FIX] Scaled metrics should be computed against correct denominator

### PR DESCRIPTION
**Pull request by elephaint**
When computing metrics on a prediction dataframe from a cross-validation output, the metrics need to be calculated against the correct cutoff. Specifically, scaled metrics (those using a `train_df`), should only "see" insample data up to the cutoff date. This PR adds the possibility to group metrics by `[id_col, cutoff_col]`, allowing correct cross-validation metrics when providing a prediction dataframe that has a `cutoff` column in it.

Todo:

- [x] Fix MQLoss
- [x] Fix scaled MQLoss
- [x] Fix calibration
- [x] Fix crps
- [x] Test all losses against previous losses - we don't actually validate output correctness! We only validate that pandas == polars, but that's trivial now
- [x] Add test case for all losses when doing cross-validation (adding `cutoff` column)
- [x] Fix spis
- [x] Fix mase
- [x] Fix rmae
- [x] Fix msse
- [x] Adds _create_train_with_cutoffs for adding cutoffs to train_df

### CODE REVIEW

File: tests/test_evaluation.py [hunk 1]
```diff
@@ -188,3 +299,482 @@ def test_distributed_evaluate(setup_series):
...
+    # Verify evaluation has correct structure
+    assert "cutoff" in eval_nw.columns, "evaluate(): cutoff column missing"
+    assert "metric" in eval_nw.columns, "evaluate(): metric column missing"
+
+    # Should have one row per (cutoff, unique_id, metric) combination
+    n_metrics = 5
+    expected_rows_eval = n_cutoffs * n_series * n_metrics
+    assert eval_nw.shape[0] == expected_rows_eval, (
+        f"evaluate(): Expected {expected_rows_eval} rows, got {eval_nw.shape[0]}"
+    )
```
Comment (elephaint) [inline tests/test_evaluation.py]: Can you add the test case where we test:

1. not using the cutoff col, i.e. dropping it and taking the mean over all the series
2. using the cutoff col and grouping afterwards (i.e. taking the mean)

has the same outcome?

File: tests/test_evaluation.py [hunk 2]
```diff
@@ -188,3 +299,482 @@ def test_distributed_evaluate(setup_series):
...
+            assert result_nw.shape[0] == expected_rows, f"calibration: Expected {expected_rows} rows"
+            assert "cutoff" in result_nw.columns, "calibration: cutoff column missing"
+
+    # Test mqloss
+    quantiles = np.array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])
+    # Create model dict for mqloss (maps model names to lists of quantile predictions)
+    model_dict = {}
+    for m in models:
+        # For simplicity, use the point forecast for all quantiles
+        model_dict[m] = [m] * len(quantiles)
```
Comment (elephaint) [inline tests/test_evaluation.py]: I'd apply a rand here to assure some variation

File: tests/test_evaluation.py [hunk 3]
```diff
@@ -188,3 +299,482 @@ def test_distributed_evaluate(setup_series):
...
+                target_col="y",
+                cutoff_col="cutoff"
+            )
+
+            result_nw = nw.from_native(result)
+            assert result_nw.shape[0] == expected_rows, f"calibration: Expected {expected_rows} rows"
+            assert "cutoff" in result_nw.columns, "calibration: cutoff column missing"
+
+    # Test mqloss
+    quantiles = np.array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])
```
Comment (elephaint) [inline tests/test_evaluation.py]: Can you test edge case where quantiles are unsorted gives the same result as when quantiles are correctly sorted? (because I think we're assuming sorted quantiles in mqloss, and we sort them in `evaluate`, but not in `mqloss`, so we might need to add an explicit sort in `mqloss` too for this edge case where people call mqloss directly with a set of quantiles.

File: tests/test_evaluation.py [hunk 4]
```diff
@@ -188,3 +299,482 @@ def test_distributed_evaluate(setup_series):
...
+    # Verify evaluation has correct structure
+    assert "cutoff" in eval_nw.columns, "evaluate(): cutoff column missing"
+    assert "metric" in eval_nw.columns, "evaluate(): metric column missing"
+
+    # Should have rows for each metric-quantile combination
+    assert eval_nw.shape[0] > 0, "evaluate(): No results returned"
+
+
+@pytest.mark.parametrize("engine", ["pandas", "polars"])
+def test_msse_rmsse_baseline_with_cutoff(engine):
```
Comment (elephaint) [inline tests/test_evaluation.py]: What does this one add compared to the scaled loss tests above?

File: utilsforecast/losses.py [hunk 5]
```diff
@@ -77,12 +87,17 @@ def _scale_loss(
     models: List[str],
     scales: IntoDataFrameT,
     id_col: str,
+    cutoff_col: str,
 ) -> IntoDataFrameT:
     exprs = [(nw.col(m) / nw.col("scale")).alias(m) for m in models]
+    group_cols = _get_group_cols(df=df, id_col=id_col, cutoff_col=cutoff_col)
+    # Determine join columns based on what's in scales dataframe
+    # (scales from train_df won't have cutoff, but scales from predictions will)
```
Comment (elephaint) [inline utilsforecast/losses.py]: I don't understand this? scales from train_df should have cutoff, because we have different train_df folds depending on the cutoff.

**Review by elephaint — COMMENTED** [review]
This is great @nasaul, thanks! I think mostly done, maybe I misunderstood but good to check whether the scaled losses now work correctly.

File: tests/test_evaluation.py (hunk 4 above)
Comment (nasaul) [reply tests/test_evaluation.py]: There was a laggy behaviour and it was fixed so it's not necessary. Removed it

File: tests/test_evaluation.py [hunk 6]
```diff
@@ -188,3 +299,482 @@ def test_distributed_evaluate(setup_series):
             distr_res.sort_values(["unique_id", "metric"]).reset_index(drop=True),
             check_dtype=False,
         )
+
+
+# ========================================
```
[the remark this answers is not in the document]
Comment (nasaul) [reply tests/test_evaluation.py]: Already done with the basic losses. Scaled losses is still missing.

[the remark this answers is not in the document]
Comment (nasaul) [reply tests/test_evaluation.py]: I've included the scaled metrics in the same test since the issue was solved in the way each loss was handled
